src/ZkFold/Symbolic/Base/Num.hs

```diff
@@ -0,0 +1,743 @@
+{-# LANGUAGE AllowAmbiguousTypes     #-}
+{-# LANGUAGE DerivingStrategies      #-}
+{-# LANGUAGE MagicHash               #-}
+{-# LANGUAGE TypeOperators           #-}
+{-# LANGUAGE UndecidableInstances    #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
+
+module ZkFold.Symbolic.Base.Num
+  ( -- * Numeric types
+    Integer
+  , Natural
+  , Rational
+  , Int
+  , Mod
+    -- * Arithmetic constraints
+  , Symbolic
+  , PrimeField
+    -- * Algebraic constraints
+  , AdditiveMonoid (..)
+  , MultiplicativeMonoid (..)
+  , AdditiveGroup (..)
+  , MultiplicativeGroup (..)
+  , Semiring
+  , Ring
+  , Field
+  , Algebra
+  , SemiEuclidean (..)
+  , Euclidean (..)
+  , SemiIntegral
+  , Integral
+  , Modular
+  , Discrete (..)
+  , Comparable (..)
+    -- * Algebraic inter-constraints
+  , From (..)
+  , Into (..)
+  , Exponent (..), (^), (^^)
+  , Scalar (..)
+    -- * Type level numbers
+  , KnownNat
+  , Prime
+  , Finite (..)
+  , FiniteChr (..)
+    -- * Numeric combinators
+  , sum
+  , product
+  , even
+  , odd
+  , fromInteger
+  , fromSemiIntegral
+  , knownNat
+  , order
+  , numberOfBits
+  , characteristic
+  , combineN
+  , combineZ
+  , evalMonoN
+  , evalMonoZ
+  ) where
+
+import           Control.Applicative
+import           Control.Category
+import           Data.Bool
+import           Data.Eq
+import           Data.Foldable       hiding (product, sum, toList)
+import           Data.Functor
+import           Data.Kind
+import           Data.Ord
+import           Data.Ratio
+import           Data.Type.Bool
+import           Data.Type.Equality
+import           GHC.Exts            (proxy#)
+import           GHC.TypeLits        (ErrorMessage (..), TypeError)
+import           GHC.TypeNats        hiding (Mod)
+import qualified GHC.TypeNats        as Type
+import           Prelude             (Int, Integer)
+import qualified Prelude
+
+-- Symbolic field extensions [Arithmetic a : a] should include:
+-- PrimeField x => Symbolic x
+-- PrimeField x => Symbolic (i -> x)
+-- PrimeField x => Symbolic (Circuit x i Par1)
+class
+  ( Field a
+  , Comparable a
+  , FiniteChr a
+  , 3 <= Chr a
+  , PrimeField x
+  , Algebra x a
+  , Chr a ~ Order x
+  ) => Symbolic x a | a -> x where
+
+-- Prime fields should only include:
+-- (SemiIntegral int, Prime p) => PrimeField (int `Mod` p)
+-- and newtypes of int `Mod` p.
+-- p = 2 is ruled out to allow/require trichotomy.
+type PrimeField x =
+  ( Modular x
+  , Finite x
+  , Symbolic x x
+  )
+
+class AdditiveMonoid a where
```

Comment:
I agree that we might not need the `AdditiveSemigroup` (and `MultiplicativeSemigroup`) base classes as we currently have no usecase for them (and all the datatypes are at least monoids)